### ISSUE TITLE
Datasets gated with start_step deadlock DDP at activation (NCCL collective timeout)

### ISSUE THREAD

**Issue by bghira** [bghira/SimpleTuner#3099]
### Symptom

On an 8-GPU DDP run using per-dataset `train_batch_size` plus `start_step` gating (three webshart image groupings: ungated @ batch 8, `start_step: 1000` @ batch 3, `start_step: 12000` @ batch 1), training proceeds normally until the first gated dataset's activation boundary, then hangs and dies at exactly that step:

```
Steps: 998/15000 ...
[rank1]:[E ... ProcessGroupNCCL.cpp:689] [Rank 1] Watchdog caught collective operation timeout:
WorkNCCL(SeqNum=34036, OpType=ALLGATHER, NumelIn=8, NumelOut=64, Timeout(ms)=5400000)
ran for 5400073 milliseconds before timing out.
  traceback : Signal 6 (SIGABRT) received by PID ...
```

One or more ranks appear to block in the newly-activated backend's first fetch while the remaining ranks reach the training step's ALLGATHER (the anyflow distiller gathers per-rank sigma/branch state each step, but any per-step collective would expose the same mismatch). After the 90-minute NCCL watchdog, the job aborts. Reproduced once at `start_step: 1000`; the run had trained 998 steps flawlessly on the ungated dataset beforehand.

### Suspected cause

Mid-run dataset activation takes a rank-asymmetric code path (iterator construction, metadata load, or a `main_process_first`-style section) that either performs a collective some ranks skip, or blocks a subset of ranks long enough that the step collective times out. Launch-time multi-dataset setup does not exhibit this — the same three-grouping config with all datasets active from step 0 trains fine.

### Workaround

Remove `start_step` gating and activate all datasets at launch; emulate curriculum phases by restarting from a checkpoint with a modified dataloader config at each phase boundary.

### Suggested fix directions

- Make gated-dataset activation collective-symmetric: pre-build the gated dataset's iterator/metadata on all ranks at startup (it is already configured and precached then), so activation is just "start drawing from it".
- Alternatively, synchronize activation behind an explicit barrier placed before the step's first collective, and log activation per rank so a hang is attributable.

**Comment by bghira** [comment]
Update — root cause reattribution after a second occurrence.

The deadlock reproduced at the same step (~998) on a config with **no `start_step` gating at all** (both datasets active from launch), so gating is not the trigger. This time the watchdog exposed the actual mechanism — two ranks entered the *same* collective with different shapes:

```
[rank1] WorkNCCL(SeqNum=34027, OpType=ALLGATHER, NumelIn=8, NumelOut=64, ...) timed out
[rank3] WorkNCCL(SeqNum=34027, OpType=ALLGATHER, NumelIn=3, NumelOut=24, ...) timed out
```

`8` and `3` are the two datasets' per-dataset `train_batch_size` values: on that step, ranks disagreed about which dataset the batch came from, and any per-sample-shaped collective (here the anyflow distiller's sigma gather; validation gathers would hit it too) mismatches and hangs. The per-step dataset selection is presumably synchronized by shared RNG, and something around step ~998 (per-rank epoch/bucket rollover with non-divisible per-rank sample counts, oversubscription padding differing per rank, etc.) shifts one rank's selection sequence by one draw.

So the bug is: **multi-dataset training with heterogeneous `train_batch_size` values is unsafe under DDP whenever any collective is batch-shaped** — a single selection desync turns into a 90-minute NCCL hang instead of an error.

Workaround that holds: keep per-dataset batch sizes equal across concurrently-active datasets; then a selection desync produces matched collective shapes (only mixing resolutions across ranks for a step) instead of a deadlock.

Suggested fixes:
- Synchronize the per-step dataset choice explicitly (broadcast the chosen backend id from rank 0 each step) instead of relying on RNG lockstep surviving per-rank epoch boundaries.
- Or detect heterogeneous batch sizes at startup and either pad batch-shaped collectives to the max batch or fail fast with a clear message.

The original `start_step` report above is likely the same desync arriving via a different path (prefetch crossing the activation boundary can shift selection on a subset of ranks); fixing the selection synchronization should address both.

**Comment by bghira** [comment]
Correction to my earlier update: the "recurrence without gating" was illusory. The persisted bucket-metadata override (#3100) had silently restored `start_step: 1000` (and the old heterogeneous batch sizes) from an earlier leg, so all three hangs — including the ones attributed to RNG-desynced dataset selection — were this same activation deadlock at the same gate. After patching the *persisted* configs to remove the gate, the mismatched-collective-shape signature and the hang should both disappear; will confirm.

Two things remain worth fixing from the investigation even so:

1. The activation deadlock itself (original report) — real, reproduced 3×, all ranks spin in a batch-shaped ALLGATHER while at least one rank blocks in the newly-activated backend's first fetch. Note the hang lands ~2 steps *before* the configured `start_step` because dataloader prefetch evaluates eligibility ahead of the optimizer step.
2. `select_dataloader_index` draws the per-step backend via `torch.multinomial(weights, 1)` on the **global CPU RNG with no dedicated generator**. Cross-rank agreement therefore depends on every rank consuming global RNG identically forever; any rank-asymmetric consumer (validation on the main rank, on-demand cache misses, dropout) desynchronizes dataset selection permanently, which turns into mismatched or missing collectives downstream. A seeded `torch.Generator` keyed on the step (or a rank-0 broadcast of the choice) would make selection deterministic regardless of RNG history.
